Mathlib/Data/DFinsupp/Multilinear.lean

```diff
@@ -0,0 +1,149 @@
+/-
+Copyright (c) 2024 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser, Sophie Morel
+-/
+import Mathlib.LinearAlgebra.DFinsupp
+import Mathlib.LinearAlgebra.Multilinear.Basic
+import Mathlib.Data.DFinsupp.Basic
+
+/-!
+# Interactions between finitely-supported functions and multilinear maps
+
+This file provides `DFinsupp.piMultilinear`, which satisfies
+`piMultilinear f x p = f p (fun i => x i (p i))`.
+
+This is useful because all the intermediate results are bundled:
+
+* `piMultilinear f x` is a `DFinsupp` supported by families of indices `p`.
+* `piMultilinear f` is a `MultilinearMap` operating on finitely-supported functions `x`.
+* `piMultilinearₗ` is a `LinearMap`, linear in the family of multilinear maps `f`.
+
+-/
+
+universe uι uκ uS uR uM uN
+variable {ι : Type uι} {κ : ι → Type uκ}
+variable {S : Type uS} {R : Type uR} {M : ∀ i, κ i → Type uM} {N : (Π i, κ i) → Type uN}
+
+namespace DFinsupp
+
+section Semiring
+
+variable [DecidableEq ι] [Fintype ι] [Semiring R]
+variable [∀ i k, AddCommMonoid (M i k)] [∀ p, AddCommMonoid (N p)]
+variable [∀ i k, Module R (M i k)] [∀ p, Module R (N p)]
+
+omit [Fintype ι] in
+/-- This lemma is sufficiently impossible to name that it's easier to just keep it private. -/
+private theorem update_aux (m : (i : ι) → Π₀ (j : κ i), M i j)
+    (i : ι) (p : (i : ι) → κ i) (x : Π₀ (j : κ i), M i j) :
+    (fun k ↦ Function.update m i x k (p k)) =
+      Function.update (fun i ↦ m i (p i)) i (x (p i)) := by
```

Comment:
Is it a dependent version of `Function.update_apply`?

Reply:
See the golf I just pushed. It's `apply_update`.

Reply:
In fact, it works inline, so I've dropped this lemma.